nntrainer/layers/rnn.cpp

```diff
@@ -95,7 +96,38 @@ void RNNLayer::setProperty(const PropertyType type, const std::string &value) {
   }
 }
 
-void RNNLayer::forwarding(bool training) {}
+void RNNLayer::forwarding(bool training) {
+  Tensor &weight_xh =
+    weightAt(static_cast<int>(RNNParams::weight_xh)).getVariableRef();
+  Tensor &weight_hh =
+    weightAt(static_cast<int>(RNNParams::weight_hh)).getVariableRef();
+  Tensor &bias_h =
+    weightAt(static_cast<int>(RNNParams::bias_h)).getVariableRef();
+
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
```

Comment:
why use loops? Simply reshape to `batch * seq_length, 1, 1, input_dim[0].width()` and dot product with the weight.
Current code creates a lot of temporary memory allocations and is complicated making it prone to errors.

Reply:
I'm not thinking about any optimization here. at least this PR. We do not have any feature about the RNN. It will update one by one.

Reply:
The current code is quite difficult to understand. This can be updated to be written functionally within 3-4 lines which are quite easy to reason about and review.

Reply:
Ok. I will try to make it easier then. Besides, I think loop in batch is necessary because of update the hidden vector between batch.

Reply:
You can refer to the FC layer implementation or use FC layer inside RNN as well :)
The operation performed here is exactly the same as FC layer with an extra dimension (which can be merged with batchsize for simplicity).

Reply:
> You can refer to the FC layer implementation or use FC layer inside RNN as well :)
> The operation performed here is exactly the same as FC layer with an extra dimension (which can be merged with batchsize for simplicity).

I think there is more than that. This is updated.